### PR TITLE
br: error if the log restore has no full backup schema or id maps

### LINKED ISSUE #54418: Performing full restore and log restore separately will be a disaster
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. Start a Log Restore.
2. Create a table named as `test.t1`.
3. Full Backup with backup ts `T1`.
4. Insert one row into the cluster.
5. Wait for log backup task checkpoint to advance.
6. Restart a new cluster
7. restore full
8. restore log restore with start ts `T1`.

### 2. What did you expect to see? (Required)
There is one row in the table `test.t1`.
### 3. What did you see instead (Required)
There is no row in the table `test.t1`.
### 4. What is your TiDB version? (Required)
v6.5, v7.1, v7.5, v8.1

### PR DESCRIPTION
**Pull request by Leavrth**
### What problem does this PR solve?

Issue Number: close #54418 

Problem Summary:
Performing full restore and log restore separately will be a disaster
### What changed and how does it work?
output error if the log restore has no full backup schema or id maps
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Prevent the cluster from performing a PITR recovery for the first time without specifying --full-backup-storage.
```

### CODE REVIEW

**Comment by kennytm** [comment]
Hold on, will this make **Full** -> **Incremental** -> **PITR** (supply `--start-ts` only) impossible to run (assume incremental has no DDLs, and ignore the RewriteTS issue at the moment)

**Comment by Leavrth** [comment]
> Hold on, will this make **Full** -> **Incremental** -> **PITR** (supply `--start-ts` only) impossible to run (assume incremental has no DDLs, and ignore the RewriteTS issue at the moment)

Yes, this PR prohibits the behavior even if there is no rewrite ts problem. The issue https://github.com/pingcap/tidb/issues/54418 provides an example.

**Comment by kennytm** [comment]
@Leavrth is the `--start-ts` flag still usable?

**Comment by Leavrth** [comment]
> @Leavrth is the `--start-ts` flag still usable?
> 
> (anyway i'm tagging `release-note` this is behavior-changing.)

Yes. Currently log restore will upload a file to the external storage, which contains `id-maps` from the upstream table/db ids to the pre-allocated downstream table/db ids.
This path of `id-maps` is: https://github.com/pingcap/tidb/blob/7a0611ff354dfbdc4085bab1d6650b77e87dbc7e/br/pkg/metautil/metafile.go#L62

With the `id-maps`, 
1. log restore can retry with checkpoint if it failed.
2. if log restore with `--restored-ts T1` is finished. The next log restore with `start-ts T1` but without `full-backup-storage` will read the `id-maps  pitr_id_maps/pitr_id_map.cluster_id:{cluster-id}.restored_ts:{start-ts}` instead of read the tableinfos in backupmeta.
https://github.com/pingcap/tidb/blob/7a0611ff354dfbdc4085bab1d6650b77e87dbc7e/br/pkg/restore/log_client/client.go#L674-L693

**Comment by kennytm** [comment]
it seems id-map was introduced in v7.1.0, so `--start-ts` requires an archive made from v7.1 at least I guess.

**Comment by Leavrth** [comment]
> it seems id-map was introduced in v7.1.0, so `--start-ts` requires an archive made from v7.1 at least I guess.

In v6.5, BR will upload a new `backupmeta_restored-ts` with latest table information after log restore done, but I have not tested it. 
https://github.com/pingcap/tidb/blob/e07d004a926ff46c654e15dbf940d770380e48c2/br/pkg/task/stream.go#L1310-L1312

**Review by 3pointer — APPROVED** [review]

File: br/pkg/restore/log_client/client.go [hunk 1]
```diff
@@ -637,15 +638,69 @@ type FullBackupStorageConfig struct {
...
 	FullBackupStorage *FullBackupStorageConfig
 }
 
+func (rc *LogClient) generateDBReplacesFromFullBackupStorage(
+	ctx context.Context,
+	cfg *InitSchemaConfig,
+) (map[stream.UpstreamID]*stream.DBReplace, error) {
+	dbReplaces := make(map[stream.UpstreamID]*stream.DBReplace)
+	if cfg.FullBackupStorage == nil {
+		envVal, ok := os.LookupEnv("UNSAFE_PITR_LOG_RESTORE_START_BEFORE_ANY_UPSTREAM_USER_DDL")
```
[the remark this answers is not in the document]
Comment (Leavrth) [reply br/pkg/restore/log_client/client.go]: done

**Review by YuJuncen — APPROVED** [review]

**Pull request merged**